cpp/src/arrow/compute/function_internal.h

```diff
@@ -0,0 +1,565 @@
+    if (!holder.is_valid) return Status::Invalid("Got null scalar");
+    return holder.value->ToString();
+  }
+};
+template <>
+struct PropertyFromScalar<std::shared_ptr<DataType>> {
+  static Result<std::shared_ptr<DataType>> FromScalar(
+      const std::shared_ptr<Scalar>& value) {
+    return value->type;
+  }
+};
+template <>
+struct PropertyFromScalar<std::shared_ptr<Scalar>> {
+  static Result<std::shared_ptr<Scalar>> FromScalar(
+      const std::shared_ptr<Scalar>& value) {
+    return value;
+  }
+};
+template <>
+struct PropertyFromScalar<std::shared_ptr<const KeyValueMetadata>> {
+  static Result<std::shared_ptr<const KeyValueMetadata>> FromScalar(
+      const std::shared_ptr<Scalar>& value) {
+    ARROW_ASSIGN_OR_RAISE(
+        const auto ty,
+        PropertyToScalar<std::shared_ptr<const KeyValueMetadata>>::GetTypeSingleton());
+    if (!value->type->Equals(ty)) {
+      return Status::Invalid("Expected ", ty->ToString(), " but got ",
+                             value->type->ToString());
+    }
+    const auto& holder = checked_cast<const MapScalar&>(*value);
+    std::vector<std::string> keys;
+    std::vector<std::string> values;
+    const auto& list = checked_cast<const StructArray&>(*holder.value);
+    const auto& key_arr = checked_cast<const BinaryArray&>(*list.field(0));
+    const auto& value_arr = checked_cast<const BinaryArray&>(*list.field(1));
+    for (int64_t i = 0; i < list.length(); i++) {
+      keys.push_back(key_arr.GetString(i));
+      values.push_back(value_arr.GetString(i));
+    }
+    return key_value_metadata(std::move(keys), std::move(values));
+  }
+};
+template <>
+struct PropertyFromScalar<Datum> {
+  static Result<Datum> FromScalar(const std::shared_ptr<Scalar>& value) {
+    if (value->type->id() == Type::LIST) {
+      const auto& holder = checked_cast<const BaseListScalar&>(*value);
+      return holder.value;
+    }
+    // TODO(ARROW-9434): handle other possible datum kinds by looking for a union
+    return Status::Invalid("Cannot deserialize Datum from ", value->ToString());
+  }
+};
+
+template <typename T>
+struct PropertyFromScalar<std::vector<T>> {
+  static Result<std::vector<T>> FromScalar(const std::shared_ptr<Scalar>& value) {
+    if (value->type->id() != Type::LIST) {
+      return Status::Invalid("Expected type LIST but got ", value->type->ToString());
+    }
+    const auto& holder = checked_cast<const BaseListScalar&>(*value);
+    if (!holder.is_valid) return Status::Invalid("Got null scalar");
+    std::vector<T> result;
+    for (int i = 0; i < holder.value->length(); i++) {
+      ARROW_ASSIGN_OR_RAISE(auto scalar, holder.value->GetScalar(i));
+      ARROW_ASSIGN_OR_RAISE(auto v, PropertyFromScalar<T>::FromScalar(scalar));
+      result.push_back(std::move(v));
+    }
+    return result;
+  }
+};
+
+// Helper to quickly define serialization for a C-style enum
+#define MAKE_ENUM_PROPERTY(ENUM, TYPE, MIN, MAX)                                        \
```

Comment:
Instead, let's add some enum utilities to `reflection_internal.h`:

```c++
// reflection_internal.h
template <typename Enum>
struct EnumTraits;

template <typename Enum,
          typename CType = typename std::underlying_type<Enum>::type>
Status ValidateEnumValue(CType raw) {
  for (auto valid : EnumTraits<Enum>::kValues) {
    if (raw == static_cast<CType>(valid)) return Status::OK();
  }
  return Status::Invalid("Invalid value for ", EnumTraits<Enum>::kName, ": ", raw);
}

template <typename Enum, Enum... E>
struct BasicEnumTraits {
  static const std::array<Enum, sizeof...(E)> kValues = {E...};
};

template <typename Enum,
          typename CType = typename std::underlying_type<Enum>::type,
          typename Type = typename CTypeTraits<T>::ArrowType>
static std::shared_ptr<DataType> UnderlyingType() {
  return TypeTraits<Type>::type_singleton();
}

template <>
struct EnumTraits<JoinOptions::NullHandlingBehavior>
    : BasicEnumTraits<JoinOptions::NullHandlingBehavior,
                          JoinOptions::NullHandlingBehavior::EMIT_NULL,
                          JoinOptions::NullHandlingBehavior::SKIP,
                          JoinOptions::NullHandlingBehavior::REPLACE> {
  static constexpr char kName[] = "JoinOptions::NullHandlingBehavior";
};
```

Reply:
This leads to some more boilerplate because 1) kValues needs to be constexpr to be initialized like so and 2) then kValues and kName need separate declarations outside the struct or the linker will complain about a missing symbol. (So every specialization needs a `static constexpr std::array<...> BasicEnumTraits<..., ..., ...>::kValues;`).